src/RHSPlibWorker.h

```diff
@@ -116,21 +118,25 @@ class RHSPlibWorker : public Napi::AsyncWorker, public RHSPlibWorkerBase {
    */
   void Execute() override {
     std::scoped_lock<std::mutex> lock{m_mutex};
-    workFunction(resultCode, returnData);
+    workFunction(resultCode, returnData, nackCode);
   }
 
   /**
-   * @brief Run the callback function to prepare the data to be sent to
-   * javascript then resolve the promise with an object that contains the
-   * `resultCode` and javascript-ready value.
+   * @brief If the result code is non-negative (no error), run the call back
+   * function and resolve the promise on the return value. Otherwise, reject the
+   * promise with the associated error.
    */
   void OnOK() override {
-    Napi::Object resultObj = Napi::Object::New(Env());
-    resultObj.Set("resultCode", resultCode);
     if (resultCode >= 0 && callbackFunction) {
-      resultObj.Set("value", callbackFunction(Env(), resultCode, returnData));
+      deferred.Resolve(callbackFunction(Env(), resultCode, returnData));
+    } else {
+      Napi::Object errorObj = Napi::Object::New(Env());
+      errorObj.Set("errorCode", resultCode);
+      if (resultCode == RHSPLIB_ERROR_NACK_RECEIVED) {
+        errorObj.Set("nackCode", nackCode);
+      }
+      deferred.Reject(errorObj);
```

Comment:
While this is a lot better than how it was, it still means that TypeScript users have no way to check read the error codes in a type-safe manner (because the `Error` type does not have `errorCode` or `nackCode` fields).

Rather than creating an error in the C++ code, we should call `Reject()` with a regular Javascript object containing the `errorCode` field (and the `nackCode` field, if applicable).

In `binding.ts`, rather than exporting `addon.Serial` and `addon.RevHub` directly, we should be exporting factory functions that instantiate the addon-defined classes, and bind a [Proxy](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Proxy/Proxy) to the newly-instantiated object. That proxy can then use [handler.apply()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Proxy/Proxy/apply) to intercept all function calls, catch our `errorCode`/`nackCode` JS object, and throw an appropriate custom `Error` subclass.

Reply:
That being said, that can wait until after you get back.